allennlp/data/dataset_readers/dataset_reader.py

```diff
@@ -176,16 +184,25 @@ def read(self, file_path: str) -> Dataset:
         else:
             cache_file = None
 
+        from torch import distributed
+
         if lazy:
+            read_fn = self._read
+            if self.max_instances is not None:
+                # Double lambda ensures that read_fn doesn't call itself recursively.
+                read_fn = lambda f: (
```

Comment:
You can't use lambda functions with pickle, so this doesn't look like it'll work in the distributed case, which shares objects via pickling.  You have to make this a class method if you want to do this redirection.  That should also make it so it's much easier to understand, also, because you can just call `self._read` in that function, instead of passing functions around.

Same comment on the one below.  You just need to add two class functions instead of this, and set `read_fn = self._read` in the default case, and `read_fn = self._something_else` in the other cases.  Though you might want to just use one function for the other cases, having it check `self.max_instances` on its own; the alternative again requires passing curried functions or lambdas around, which won't work with pickle.

Reply:
This does work. They don't need to be pickled as they are instantiated in the workers. But I'm not sure this is the most readable way of wrapping these functions.

Reply:
Are you certain?  Have you tried lazy + distributed together?  Passing a lambda function to `LazyInstances` below was precisely the cause of a problem that we fixed recently: https://github.com/allenai/allennlp/pull/4026.

Reply:
Actually, I think you are right. This only worked in my experiments because I wasn't trying a lazy dataset.

Reply:
50 lines of code later, this now also works with lazy datasets.